OpenRA.Mods.D2k/Traits/World/VariedCostManager.cs

```diff
@@ -0,0 +1,53 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	[Desc("Attach this to the world actor.")]
+	public class VariedCostManagerInfo : ITraitInfo
+	{
+		[Desc("Interval between new pricings measured in ticks.")]
+		public readonly int AdjustmentDelay = 1000;
+
+		[Desc("Percentage modifier to apply.")]
+		public readonly int2 Multiplier = new int2(75, 100);
+
+		public object Create(ActorInitializer init) { return new VariedCostManager(this); }
+	}
+
+	public class VariedCostManager : ITick, ISync
+	{
+		readonly VariedCostManagerInfo info;
+
+		[Sync]
+		int cachedCostPercentage;
```

Comment:
This should be calculated seperately for each actor.

But i would rather the combat tanks to share the percentage while being seperate actors. This was an issue i had in my initial implementation with it.

Reply:
This is a world trait not a per actor trait.

Reply:
I'm aware, but the modifiers shouldn't be same for each actor at any given time. It should be calculated and saved per actor, or some sort of class that's defined on the actor trait so combat tanks can share the modifier between each other.

Reply:
I am afraid that this is technically impossible as `IProductionCostModifierInfo` lies on the `TraitInfo`. If I let all `VariedCostMultiplier` tick and define random prizes on the individual actors it won't help as `IProductionCostModifierInfo´ is not able to access them.

Reply:
You can add a Class to the Modifier. Then change CostPercentage to a Pair<String, int>[]. Add an initialisaion for this pair array that checks all actors in the mod with the Modifier trait, with Class (or actor codename if Class not defined) for String and 100 as the int. Then do the random modifier calculation in a foreach for each Class/Actor in that array. Then read the matching thing from the array from the IProductionCostMultiplierInfo.

Reply:
Okay, that is indeed a slightly better simulation of a real world market where changes apply to the individual product. However we lost `ISync` support.